### PR TITLE
fix(gatsby): Reset graphql runner on certain node events

### LINKED ISSUE #18661: Hot reloading error: "path" argument must be of type string
Hot reloading in `gatsby develop` gives this error in my project. Hot reloading sometimes works despite the error (as in, I see the error in the console window where I ran `gatsby develop`, but the website in browser appears to hot reload and work just fine). This error appeared after I updated some (mainly Gatsby) npm dependencies. Stack trace is giving me no ideas on how to debug this. Is it a Gatsby bug?

Note that this error only comes up on hot reload. So I can run `gatsby develop` just fine as long as I don't make any changes to any files. I can also build just fine.

```
[ { TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Received type undefined
      at assertPath (path.js:39:11)
      at Object.resolve (path.js:1085:7)
      at findLinkedFileNode (/.../node_modules/gatsby/dist/schema/resolvers.js:208:47)
      at resolveValue (/.../node_modules/gatsby/dist/schema/resolvers.js:223:108)
      at /.../node_modules/gatsby/dist/schema/resolvers.js:220:10
      at process._tickCallback (internal/process/next_tick.js:68:7)
    message:
     'The "path" argument must be of type string. Received type undefined',
    locations: [ [Object] ],
    path:
     [ 'allMarkdownRemark', 'edges', 0, 'node', 'frontmatter', 'cover' ] } ]

```

### PR DESCRIPTION
**Pull request by sdobz**
Hot reloading is broken on certain configurations due to node materialization.
Child objects in the node model are not updated when a node is recreated
This has something to do with the recent materialization update

By resetting the runner on these events the node model is forced to be recreated

See details in issue #18661, this is a dirty fix but allows me to edit files without
gatsby crashing

## Description

Resets the GraphQLRunner in `packages/gatsby/src/bootstrap/graphql-runner.js` on events
* `DELETE_CACHE`,
* `CREATE_NODE`,
* `DELETE_NODE`,
* `DELETE_NODES`,
* `SET_SCHEMA_COMPOSER`,
* `SET_SCHEMA`,
* `ADD_FIELD_TO_NODE`,
* `ADD_CHILD_NODE_TO_PARENT_NODE`,

### Documentation

This is a direct translation of comments by @pieh [here](https://github.com/gatsbyjs/gatsby/issues/18661#issuecomment-546893796)

## Related Issues
Fixes #18661

### CODE REVIEW

**Comment by sdobz** [comment]
As mentioned in the issue above this is probably not the right fix and no upset will be taken if rejected.
I am just trying to highlight this issue and present a quick fix for anyone else experiencing this issue

**Comment by rosnovsky** [comment]
You're totally onto something here! It fixed the error I've been getting, and `gatsby develop` no longer breaks whenever I save a file. The _hot-reload_ doesn't happen and I have to hit refresh manually, but at least I don't have to restart `gatsby develop` every time I want a preview. Thank you so much!

File: packages/gatsby/src/bootstrap/graphql-runner.js [hunk 1]
```diff
@@ -3,8 +3,24 @@ const stackTrace = require(`stack-trace`)
 const GraphQLRunner = require(`../query/graphql-runner`)
 const errorParser = require(`../query/error-parser`).default
 
+const { emitter } = require(`../redux`)
+
 module.exports = (store, reporter) => {
-  const runner = new GraphQLRunner(store)
+  let runner = new GraphQLRunner(store)
```
Comment (freiksenet) [inline packages/gatsby/src/bootstrap/graphql-runner.js]:
```suggestion
  // TODO: Move tracking of changed state inside GraphQLRunner itself. https://github.com/gatsbyjs/gatsby/issues/20941
  let runner = new GraphQLRunner(store)
```

**Comment by sdobz** [comment]
The CI failure seems attached to an intermittent NPM issue.
```
info Done building in 29.199 sec
npm ERR! code ENOAUDIT
npm ERR! audit Your configured registry (https://registry.npmjs.org/) may not support audit requests, or the audit endpoint may be temporarily unavailable.
```

I'm reluctant to push a no-op commit to trigger a CI rerun, can someone on the gatsby team investigate the CI issue?

**Comment by pieh** [comment]
> can someone on the gatsby team investigate the CI issue?

I just restarted this one job. This is often just fluke (particularly then "audit endpoint may be temporarily unavailable" seems to often be the case because restarts usually fix those)

**Comment by maxpou** [comment]
I'm able to reproduce #18661 on [my personal blog](https://github.com/maxpou/maxpou.fr).

I pulled the code locally and it fix the issue.

**Comment by vladar** [comment]
Thanks for the fix! 🥇

**Pull request merged**